src/Storages/System/StorageSystemDetachedParts.cpp

```diff
@@ -320,16 +322,33 @@ class ReadFromSystemDetachedParts : public SourceStepWithFilter
     std::shared_ptr<StorageSystemDetachedParts> storage;
     std::vector<UInt8> columns_mask;
 
-    const ActionsDAG::Node * predicate = nullptr;
+    ActionsDAGPtr filter;
     const size_t max_block_size;
     const size_t num_streams;
 };
 
 void ReadFromSystemDetachedParts::applyFilters(ActionDAGNodes added_filter_nodes)
 {
+    filter_actions_dag = ActionsDAG::buildFilterActionsDAG(added_filter_nodes.nodes);
     filter_actions_dag = ActionsDAG::buildFilterActionsDAG(added_filter_nodes.nodes);
     if (filter_actions_dag)
-        predicate = filter_actions_dag->getOutputs().at(0);
+    {
+        const auto * predicate = filter_actions_dag->getOutputs().at(0);
+
+        Block block;
+        block.insert(ColumnWithTypeAndName({}, std::make_shared<DataTypeString>(), "database"));
+        block.insert(ColumnWithTypeAndName({}, std::make_shared<DataTypeString>(), "table"));
+        block.insert(ColumnWithTypeAndName({}, std::make_shared<DataTypeString>(), "engine"));
+        block.insert(ColumnWithTypeAndName({}, std::make_shared<DataTypeUInt8>(), "active"));
+        block.insert(ColumnWithTypeAndName({}, std::make_shared<DataTypeUUID>(), "uuid"));
+
+        filter = VirtualColumnUtils::splitFilterDagForAllowedInputs(predicate, &block);
+        if (filter)
+        {
+            auto empty_block = block.cloneWithColumns(block.cloneEmptyColumns());
+            VirtualColumnUtils::filterBlockWithDAG(filter, empty_block, context);
```

Comment:
can you explain what are we doing here?

Reply:
It was needed for building sets.
I've changed this to call a separate function to make it more clear.